drivers/flash/soc_flash_mcux.c

```diff
@@ -204,36 +204,49 @@ static int flash_mcux_read(const struct device *dev, off_t offset,
 	 */
 	addr = offset + priv->pflash_block_base;
 
-#ifdef CONFIG_CHECK_BEFORE_READING
-  #ifdef CONFIG_SOC_LPC55S36
-	/* Validates the given address range is loaded in the flash hiding region. */
-	rc = FLASH_IsFlashAreaReadable(&priv->config, addr, len);
-	if (rc != kStatus_FLASH_Success) {
-		rc = -EIO;
-	} else {
-		/* Check whether the flash is erased ("len" and "addr" must be word-aligned). */
-		rc = FLASH_VerifyErase(&priv->config, ((addr + 0x3) & ~0x3),  ((len + 0x3) & ~0x3));
+#if defined(CONFIG_CHECK_BEFORE_READING) && !defined(CONFIG_SOC_LPC55S36)
```

Comment:
A suggestion here for readability- could we do something like this:

```
if (IS_ENABLED(CONFIG_CHECK_BEFORE_READING)) {
    if (IS_ENABLED(CONFIG_SOC_LPC55S36) {
        /* LPC55S36 specific logic */
    } else {
       rc = is_area_readable(addr, len)
    }
}

if (!rc) {
   memcpy(data, (void *)addr, len);
} else if ((rc == -ENODATA) && IS_ENABLED(CONFIG_CHECK_BEFORE_READING)) {
  /* Erased area, return dummy data as an erased page. */
  memset(data, 0xFF, len);
  rc = 0;
}
return rc;
```

Apologies if formatting there is confusing. At least IMO, the current function implementation in `main` is hard to follow, and I figure if we are fixing the bug cleaning up the readability wouldn't hurt either

Reply:
Completely agree, but it's really a thorny situation. The `FLASH_Read` does a "copy before verify" which doesn't neatly fit in the previous model:

| `CHECK` is set? | on `55S36`? | Check stage | Copy stage | `0xff` fill 
|--------|--------|--------|-------|-------|
| no | - | no | `memcpy` | no |
| yes| no | `is_area_readable` | `memcpy` | yes |
| yes| yes| `FLASH_Read` + `VerifyErased` | (already done) | yes

My first proposal avoided any duplicate code, but ended up creating awkward logical groupings.

Reply:
Thank you for the explanation here, this makes more sense (and I really like the rework in the latest push, it is a lot easier to follow)